### DISCUSSION TITLE
Security and rate limiting

### DISCUSSION THREAD

**Discussion by dcheung2** [Netflix/dgs-framework#31] [Q&A]
One barrier I have when onboarding GraphQL to our technology stack is security.
It is not the security about authentication or authorization / permission / ACL which can be solved by Spring.

I mean more about SRE/DoS attack protection, 
- timeout
- rate limiting
- throttling
- Query depth
- QoS

The other side, REST API, they claim that it could solved simply on an external API gateway.

[Few](https://www.howtographql.com/advanced/4-security/) [thing](https://github.com/graphql-java/graphql-java/issues/291) [I've read](https://medium.com/in-the-weeds/graphql-and-resource-limitations-442c3bd72358) and seem that [graph-java support it](https://github.com/graphql-java/graphql-java/pull/655).

Is DGS support it already ? If yes, where could I find an example of it ?

**Comment by srinivasankavitha** [comment]
One of the common techniques to limit is to check the  Query depth by instrumenting the datafetchers for each field. There is
already a `MaxQueryComplexityInstrumentation` class in `graphql-java` that
you can use to add that functionality:
https://www.javadoc.io/doc/com.graphql-java/graphql-java/10.0/graphql/analysis/MaxQueryComplexityInstrumentation.html

The DGS Framework supports adding your own instrumentation classes to add
additional checks or instrumentation. So you can add one that implements
the `MaxQueryComplexityInstrumentation` and add any additional logic that
you need. Here is an example that adds tracing:
https://netflix.github.io/dgs/advanced/instrumentation/

Hope this helps.

**Comment by paulbakker** [comment]
DGS doesn’t support this out of the box, because it’s hard to come up with something that works for different usage scenarios. The only place we need this kind of functionality such as rate limiting at Netflix is in a federated context, where the gateway is responsible for these concerns. 

I _think_ such concerns are typically better handled by a gateway before your DGS instances are even hit, but definitely open to other ideas.
